swift/ql/lib/codeql/swift/controlflow/internal/AstControlFlowTrees.qll

```diff
@@ -0,0 +1,264 @@
+private import swift
+private import Completion
+private import ControlFlowGraphImplShared
+private import ControlFlowElements
+
+module ControlFlowTrees {
+  abstract class AstControlFlowTree extends AstNode {
+    /** Holds if `first` is the first element executed within this element. */
+    pragma[nomagic]
+    abstract predicate first(ControlFlowElement first);
+
+    /**
+     * Holds if `last` with completion `c` is a potential last element executed
+     * within this element.
+     */
+    pragma[nomagic]
+    abstract predicate last(ControlFlowElement last, Completion c);
+
+    /** Holds if abnormal execution of `child` should propagate upwards. */
+    abstract predicate propagatesAbnormal(ControlFlowElement child);
+
+    /**
+     * Holds if `succ` is a control flow successor for `pred`, given that `pred`
+     * finishes with completion `c`.
+     */
+    pragma[nomagic]
+    abstract predicate succ(ControlFlowElement pred, ControlFlowElement succ, Completion c);
+  }
+
+  private class ControlFlowTreeAdaptor extends ControlFlowTree {
+    AstControlFlowTree tree;
+
+    ControlFlowTreeAdaptor() { tree = this.asAstNode() }
+
+    override predicate first(ControlFlowElement first) { tree.first(first) }
+
+    override predicate last(ControlFlowElement last, Completion c) { tree.last(last, c) }
+
+    override predicate propagatesAbnormal(ControlFlowElement child) {
+      tree.propagatesAbnormal(child)
+    }
+
+    override predicate succ(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
+      tree.succ(pred, succ, c)
+    }
+  }
+}
+
+import ControlFlowTrees
+
+/**
+ * Holds if `first` is the first element executed within control flow
+ * element `cft`.
+ */
+predicate astFirst(AstControlFlowTree cft, ControlFlowElement first) { cft.first(first) }
+
+/**
+ * Holds if `last` with completion `c` is a potential last element executed
+ * within control flow element `cft`.
+ */
+predicate astLast(AstControlFlowTree cft, ControlFlowElement last, Completion c) {
```

Comment:
Couldn't this be `last(any(ControlFlowTree tree | tree.asAstNode()), last, c`?

Reply:
Indeed, that's much better. I've pushed that change now: https://github.com/github/codeql/pull/9291/commits/1f4924f978bc5e3fff7a32c13cc9e307085a611a#diff-a3a428364dc34dde5d544b5ba69082c665a60fdf4dee850acb1dddd46e901b5bR18

To keep the commit history clean for the reviewers on the C team I've forced pushed your proposed change. That makes it slightly more difficult for you, but I think it'll make the PR a lot easier to review for everybody else.